File: cpp/62.Unique-Paths.DP.cpp
// 100% faster
// O(m+n)

class Solution {
public:
    int uniquePaths(int m, int n) {
        int table[m][n];
        table[m-1][n-1] = 1;  // I don't think it should be counted as a path though
        for(int i{n-2}; i>=0; i--){
            table[m-1][i] = 1;
        }
        for(int i{m-2}; i>=0; i--){
            table[i][n-1] = 1;
        }
        for(int i{m-2}; i>=0; i--){
            for(int j{n-2}; j>=0; j--){
                table[i][j] = table[i][j+1] + table[i+1][j];
            }
        }
        return table[0][0];
    }
};

File: Java/62. Unique Paths.DP.java
class Solution {
    public int uniquePaths(int m, int n) {
        if (m == 1 || n == 1) {
            return 1;
        }
        int[][] dp = new int[m][n];
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                // dp[0][0]
                if (i == 0 && j == 0) {
                    dp[i][j] = 0;
                    continue;
                }
                // dp[i][0] or dp[0][j]
                if (i == 0 || j == 0) {
                    dp[i][j] = 1;
                    continue;
                }
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[m - 1][n - 1];
    }
}
// time complexity : faster 100% O(m * n)
// space complexity : faster 75.15% O(m * n)
